agent/src/com/jonnyzzz/teamcity/virtual/VMDetector.java

```diff
@@ -46,44 +47,68 @@ public void beforeAgentConfigurationLoaded(@NotNull BuildAgent agent) {
 
       private void detectVagrant(@NotNull final BuildAgentConfiguration config) {
 
-        final String output = executeCommandWithShell("vagrant", "vagrant --version");
+        final String[] output = executeCommandWithShell("vagrant", "vagrant --version");
         if (output == null) return;
 
-        String ver = output.toLowerCase().trim();
+        String ver = output[0].toLowerCase().trim();
         if (ver.startsWith("vagrant")) ver = ver.substring("vagrant".length()).trim();
 
         if (StringUtil.isEmptyOrSpaces(ver)) {
-          LOG.warn("Failed to parse vagrant version: " + output);
+          LOG.warn("Failed to parse vagrant version: " + Arrays.toString(output));
           return;
         }
 
         config.addConfigurationParameter(VMConstants.VAGRANT_PROPERTY, ver);
       }
 
       private void detectDocker(@NotNull final BuildAgentConfiguration config) {
-        if (!SystemInfo.isLinux) {
-          LOG.debug("Docker is only available under Linux");
-          return;
-        }
 
-        final String output = executeCommandWithShell("docker", "docker --version");
+        final String[] output = executeCommandWithShell("docker", "docker version");
         if (output == null) return;
 
-        String ver = output.toLowerCase().trim();
-        ver = ver.replaceAll("\\s*docker\\s+version\\s+", "");
-        ver = ver.replaceAll(",?\\s+build\\s+", "-");
+        String[] cleanedArray = new String[output.length];
 
-        if (StringUtil.isEmptyOrSpaces(ver)) {
-          LOG.warn("Failed to parse docker version: " + output);
+        for (int i = 0; i < output.length; i++)
+          cleanedArray[i] = output[i].trim();
+
+        int serverPosition = Arrays.asList(cleanedArray).indexOf("Server:");
```

Comment:
Can you simply have a regex to search all necessary parameter values?

It looks like `stdout.split("Server:")` and  `stdout.split("Client:")` will be enough to select client and server blocks. And next a regex could be used to parse the rest.

We shall have the implementation as independent from command output as possible. Docker will change, and out code is likely to work without further changes